src/sisl/io/sile.py

```diff
@@ -722,72 +718,44 @@ def wrap_func(*args, **kwargs):
-                if step is None:
-                    step = 1
-            else:
-                # behave like range(stop) | range(start, stop, step)
-                # almost.
-                #  step only => start, step = 0, inf
-                #  start only => stop = start + 1
-                #  stop only => start = stop - 1
-                if step is not None:
-                    if start is None:
-                        start = 0
-                    if stop is None:
-                        stop = -1
-                elif start is None and stop is None:
-                    start = 0
-                    stop = 1
-
-                if start is None:
-                    if stop is not None:
-                        start = stop - 1
-                    else:
-                        start = 0
-                if stop is None:
-                    # this will only occur if start is given, but not stop
-                    stop = start + 1
-                elif stop < 0:
-                    stop = 1000000000000
-                if step is None:
-                    # only handle cases where the requested number of
-                    # elements is longer than 1, then step should be defined,
-                    # else we resort to returning a single entity (not postprocessed)
-                    if stop - start > 1:
-                        step = 1
-
-            if stop < 0:
-                # TODO this will probably fail if users do stop=-4 (to not have the last 3 items)
-                # It should be done in a different manner
-                stop = 100000000000
+            inf = 100000000000
 
             def check_none(r):
                 if isinstance(r, tuple):
                     return reduce(lambda x, y: x and y is None, r, True)
                 return r is None
 
-            # start by reading past start, this is regardless of what we
-            # are about to do
-            for _ in range(start):
-                r = skip_call(*args, **kwargs)
-                if check_none(r):
-                    return r
-
+            if start is None:
+                start = 0
+            if stop is None:
+                stop = inf
             if step is None:
-                # read a single element and return
-                def ret_func():
-                    return wrap_func(*args, **kwargs)
+                step = 1
 
+            if start < 0 or stop < 0 or step < 0:
+                # we need to read everything and take a slice afterwards
+                def ret_func():
+                    R = []
+                    for _ in range(inf):
+                        r = wrap_func(*args, **kwargs)
+                        if check_none(r):
+                            break
+                        R.append(r)
+                    slc = slice(start, stop, step)
+                    R = R[slc]
+                    if len(R) == 1:
+                        return R[0]
+                    return postprocess(R)
```

Comment:
I am not sure I like this.

Consider users doing something that results in only returning a single entry at a single iteration. Now all other iterations will use the `GeometryCollection` (or postprocess call) and just one of them will use the raw output.

Wouldn't it be better to force users to do `read_()[0]` in that case? I know it is a bit annoying, but...

Otherwise we could add a `single` argument that replaces `start/stop` with the one that is given.
I.e.
```python
read_geometry(start=3, single=True) == read_geometry(stop=3, single=True)
```

Reply:
I kind of like the idea of `single` (opposite effect from the original `all`), but it introduces an extra keyword that can be avoided if we can tolerate single-entry cases are handled differently than multi-entry cases.

Anyways,
> ```python
> read_geometry(start=3, single=True) == read_geometry(stop=3, single=True)
> ```
seems a bit confusing to me since with `stop=3` in a `slice` would not include the fourth element. Why not

```python
read_geometry(start=3, single=True) == read_geometry(stop=4, single=True)
```
?
How should eg. `read_geometry(start=2, stop=4, single=True)` be handled?

Reply:
Another possibility could be to let `step=0` mean first entry with `start` and last entry with `stop`.

Reply:
If `single` is true, then invalid input should result in an exception.
Yeah, probably only `start` should be valid input.

> Another possibility could be to let `step=0` mean first entry with `start` and last entry with `stop`.

I don't think so, that is somewhat counterintuitive since `stop` is generally not included in `range` expressions. If users want that, they should do `start=, stop=+1, step=stop-start`

Reply:
I meant `step=0` to behave as `single=True`, because that value is not permitted in a slice. We would avoid a new keyword.

Reply:
Ah, I see. Yeah, that could probably also work. I can agree to get this in, but perhaps we should then hold on the other things before we have discussed the commented things (see further down)